### DISCUSSION TITLE
Coverage with code optimization

### DISCUSSION THREAD

**Discussion by dirkvdmerwe** [cilium/coverbee#8] [General]
Hi @dylandreimerink, thanks for the work on Coverbee, we have found this to be a very useful tool!

I had a question regarding a comment made last year about the code coverage reported with optimized code [here](https://github.com/cilium/cilium/pull/20230#issuecomment-1182936312)

> I am working on an coverbee update that improves this by combining the results with static analysis of the source code, but that is still going to take some time.

Is this on the project roadmap? 
Also, I am curious whether folks have any suggestions to improve the reported coverage.

**Comment by dylandreimerink** [comment]
The improvements I mentioned in that comment were added with https://github.com/cilium/coverbee/pull/1. It uses this very crude C parser https://github.com/cilium/coverbee/tree/master/pkg/cparser to improve the display. 

Essentially the idea is to figure out which lines of code all execute together (form a basic block) so that if I can map a basic block to a line in the source code, then I know I can color the whole code block green even if some statements were optimized out ect. This results in a much more accurate looking result. Before this you could have situations where 1 expression would be red and the one after it green while that would be impossible in practice.

> Is this on the project roadmap?
> Also, I am curious whether folks have any suggestions to improve the reported coverage.

I personally have a few things I would like to do. Though I have not formally made issues out of them yet:
* [ ] Make the C parsing + transposing step optional, at least fall back gracefully if it doesn't work
* [ ] Make an Assembly report, so showing coverage on the BPF assembly
* [ ] Improve the C parser so it understands preprocessor statements and can map coverage back to macro values
* [ ] Improve the instrumentation to reduce overhead
* [ ] Start using global data instruction (require kernel >5.2) and implement a probe + maintain current implementation as fallback
* [ ] Output a instrumented ELF binary instead of having to load via Golang
* [ ] Figure out a way to calculate available registers / stack locations without relying on verbose verifier log parsing (its not long term stable)

So lots to do, some of these are more ambitious than others :grin: . I also welcome feedback, so let me know what you think or would like to see!

**Reply by dirkvdmerwe** [reply]
Thanks for the feedback.

Yes, I see, definitely some interesting features you are considering...
Since you are asking, from my selfish perspective :), I would like to see the BPF assembly report and the instrumented ELF binary features.

But either way, look forward to seeing what you guys come up with.
